clients/src/main/java/org/apache/kafka/clients/consumer/internals/DefaultBackgroundThreadRunnable.java

```diff
@@ -0,0 +1,235 @@
+                                           BlockingQueue<ApplicationEvent> applicationEventQueue,
+                                           BlockingQueue<BackgroundEvent> backgroundEventQueue,
+                                           SubscriptionState subscriptions,
+                                           ApiVersions apiVersions,
+                                           Metrics metrics,
+                                           ClusterResourceListeners clusterResourceListeners,
+                                           Sensor fetcherThrottleTimeSensor) {
+        try {
+            this.time = time;
+            this.log = logContext.logger(DefaultBackgroundThreadRunnable.class);
+            this.applicationEventQueue = applicationEventQueue;
+            this.backgroundEventQueue = backgroundEventQueue;
+            this.config = config;
+            setConfig();
+            this.inflightEvent = Optional.empty();
+            this.subscriptions = subscriptions;
+            this.metrics = metrics;
+            this.metadata = bootstrapMetadata(clusterResourceListeners, logContext);
+            ChannelBuilder channelBuilder = ClientUtils.createChannelBuilder(config, time, logContext);
+            NetworkClient netClient = new NetworkClient(
+                    new Selector(config.getLong(ConsumerConfig.CONNECTIONS_MAX_IDLE_MS_CONFIG), metrics, time, METRIC_GRP_PREFIX, channelBuilder, logContext),
+                    this.metadata,
+                    clientId,
+                    100, // a fixed large enough value will suffice for max in-flight requests
+                    config.getLong(ConsumerConfig.RECONNECT_BACKOFF_MS_CONFIG),
+                    config.getLong(ConsumerConfig.RECONNECT_BACKOFF_MAX_MS_CONFIG),
+                    config.getInt(ConsumerConfig.SEND_BUFFER_CONFIG),
+                    config.getInt(ConsumerConfig.RECEIVE_BUFFER_CONFIG),
+                    config.getInt(ConsumerConfig.REQUEST_TIMEOUT_MS_CONFIG),
+                    config.getLong(ConsumerConfig.SOCKET_CONNECTION_SETUP_TIMEOUT_MS_CONFIG),
+                    config.getLong(ConsumerConfig.SOCKET_CONNECTION_SETUP_TIMEOUT_MAX_MS_CONFIG),
+                    time,
+                    true,
+                    apiVersions,
+                    fetcherThrottleTimeSensor,
+                    logContext);
+            this.networkClient = new ConsumerNetworkClient(
+                    logContext,
+                    netClient,
+                    metadata,
+                    time,
+                    retryBackoffMs,
+                    config.getInt(ConsumerConfig.REQUEST_TIMEOUT_MS_CONFIG),
+                    heartbeatIntervalMs);
+        } catch (Exception e) {
+            // now propagate the exception
+            throw new KafkaException("Failed to construct background processor", e);
+        }
+    }
+
+    // VisibleForTesting
+    DefaultBackgroundThreadRunnable(Time time,
```

Comment:
Should this constructor call the "main" constructor? If not, it should call `setConfig` too, right?

Reply:
yap

Reply:
This constructor is meant for testing because a few of the clients are actually need to be initiated here locally, e.g. metadata, network client.